xformers/triton/k_dropout.py

```diff
@@ -10,138 +10,232 @@
-            inputs += b
-
-        act_grad = META["ACTIVATION_GRAD"](inputs)
-        grad_out *= act_grad
-
-    # randomly prune it
-    if p > 0.:
-        output = _drop_and_scale(SEEDS, row, p, grad_offsets, grad_out)
-    else:
-        output = grad_out
-
-    # write-back
-    y_ptrs = GRAD_IN + grad_offsets
-    tl.store(y_ptrs, output, mask=mask)
+    BLOCK_M = meta["BLOCK_M"]
+    BLOCK_N = meta["BLOCK_N"]
+    SIZE_RAND_BLOCK = meta["SIZE_RAND_BLOCK"]
+    TRAINABLE_BIAS = meta["TRAINABLE_BIAS"]
+
+    rows = tl.arange(0, BLOCK_M)
+    row_id = tl.program_id(axis=0)
+    rows = row_id * BLOCK_M * 4 + tl.arange(0, BLOCK_M)
+
+    col_id = tl.program_id(axis=1)
+    cols = col_id * BLOCK_N + tl.arange(0, BLOCK_N)
+    seed = SEEDS + col_id  # FIXME index the seed properly
+
+    # pointers starting point
+    grad_out_ptrs = GRAD_OUT + rows[:, None] * stride_grad + cols[None, :]
+    grad_in_ptrs = GRAD_IN + rows[:, None] * stride_grad + cols[None, :]
+    input_ptrs = INPUTS + rows[:, None] * stride_inputs + cols[None, :]
+
+    # random binary masks, save registers
+    rand_offsets = tl.arange(0, SIZE_RAND_BLOCK) + row_id * BLOCK_M * 4
+    rand1, rand2, rand3, rand4 = tl.randint4x(seed.to(tl.int32), rand_offsets)
+
+    # NOTE: Should have been possible to do with a int threshold instead, but seems numerically wrong
+    rand_mask1 = tl.uint32_to_uniform_float(rand1) > p
+    rand_mask2 = tl.uint32_to_uniform_float(rand2) > p
+    rand_mask3 = tl.uint32_to_uniform_float(rand3) > p
+    rand_mask4 = tl.uint32_to_uniform_float(rand4) > p
+
+    # now go over the tiles
+    grad_bias = tl.zeros((BLOCK_N,), dtype=tl.float32)
+    col_mask = cols[None, :] < N
+    zero = 0.0
+    p_scale = 1 / (1 - p) if p < 1. else 1.
+
+    if meta["USE_BIAS"]:
+        b_ptrs = BIAS + cols[None, :]
+        bias = tl.load(b_ptrs, mask=col_mask, other=0.)
+
+    for i in range(4):
+        # cycle through the binary masks (workaround / no indexing)
+        if i == 0:
+            rand_mask = rand_mask1
+        elif i == 1:
+            rand_mask = rand_mask2
+        elif i == 2:
+            rand_mask = rand_mask3
+        else:
+            rand_mask = rand_mask4
+
+        block_mask = (rows[:, None] < M) & col_mask
+        grad_out = tl.load(grad_out_ptrs, mask=block_mask, other=0.)
+
+        # optional: fused activation (while the data is in shared memory)
+        if meta["ACTIVATION_GRAD"]:
+            inputs = tl.load(input_ptrs, mask=block_mask, other=0.)
+
+            # optionally apply a fused bias
+            if meta["USE_BIAS"]:
+                inputs += bias
+
+            act_grad = meta["ACTIVATION_GRAD"](inputs).to(grad_out.dtype)
+            grad_out *= act_grad
+
+        # randomly prune and scale
+        if p > 0.:
+            # generate all the random numbers for the block at once, then reshape
+            keep = tl.reshape(rand_mask, grad_out.shape)
+
+            # prune and normalize in one go
```

Comment:
nit: seems like there's a lot of duplication with the fwd, maybe easy to just pull out into a common method?

Reply:
I'll clean that up, you're right, I'll need too kernels still but I can factorize a couple of functions

Reply:
@dianaml0 I could not change that much in the end, I just gave it a try, the triton jit phase chokes on some of the factorization options

Reply:
ok, fixed by changing some of the formulations, and small speedup on the dropmask generation at the same time. This should be covered, running one last full GPT run and landing if that's ok